### PR TITLE
Codingstyle guideline that definitions should be specified

### PR DESCRIPTION
**Pull request by EgbertRijke**
I made a coding style guideline that definitions should be specified, and specified some definitions about (co)presheaves.

### CODE REVIEW

File: src/category-theory/copresheaf-categories.lagda.md [hunk 1]
```diff
@@ -65,34 +66,98 @@ module _
...
-    right-unit-law-comp-hom-Large-Precategory copresheaf-Large-Precategory
+  comp-hom-copresheaf-Large-Category :
+    {l3 l4 l5 : Level}
+    (X : obj-copresheaf-Large-Category l3)
+    (Y : obj-copresheaf-Large-Category l4)
+    (Z : obj-copresheaf-Large-Category l5) →
+    hom-copresheaf-Large-Category Y Z → hom-copresheaf-Large-Category X Y →
+    hom-copresheaf-Large-Category X Z
+  comp-hom-copresheaf-Large-Category X Y Z g h =
+    comp-hom-Large-Precategory copresheaf-Large-Precategory {X = X} {Y} {Z} g h
```
Comment (fredrik-bakke) [inline src/category-theory/copresheaf-categories.lagda.md:96]: Is there a particular reason you want these to be explicit instead? This breaks #896.

**Review by fredrik-bakke — APPROVED** [review]
Looks good! Thanks for adding the justification to the guide. There may be one more category theory file where I took this shortcut, but I'll look through them and fix it if I find it.

File: src/category-theory/copresheaf-categories.lagda.md (hunk 1 above)
Comment (EgbertRijke) [reply src/category-theory/copresheaf-categories.lagda.md:96]: Agda couldn't infer the arguments if they were implicit.

Note that this couldn't be detected in the untyped definition, but the untyped definition also wasn't used anywhere yet, so I suspect that you would have discovered that these arguments needed to be explicit once you started using them. Is this not the case?

Comment (fredrik-bakke) [reply src/category-theory/copresheaf-categories.lagda.md:96]: Turns out I only used `hom-copresheaf-Large-Category` in that one, so I wouldn't have the experience to know yet.

File: src/category-theory/copresheaf-categories.lagda.md [hunk 2]
```diff
@@ -65,34 +66,98 @@ module _
...
+  associative-comp-hom-copresheaf-Large-Category X Y Z W h g f =
+    associative-comp-hom-Large-Precategory
+      ( copresheaf-Large-Precategory)
+      { X = X}
+      { Y}
+      { Z}
+      { W}
+      ( h)
+      ( g)
+      ( f)
```
Comment (fredrik-bakke) [inline src/category-theory/copresheaf-categories.lagda.md:128]:
```suggestion
  associative-comp-hom-copresheaf-Large-Category X Y Z W =
    associative-comp-hom-Large-Precategory
      ( copresheaf-Large-Precategory)
      { X = X}
      { Y}
      { Z}
      { W}
```

File: src/category-theory/copresheaf-categories.lagda.md (hunk 1 above)
Comment (fredrik-bakke) [inline src/category-theory/copresheaf-categories.lagda.md:96]:
```suggestion
  comp-hom-copresheaf-Large-Category X Y Z =
    comp-hom-Large-Precategory copresheaf-Large-Precategory {X = X} {Y} {Z}
```

File: src/category-theory/copresheaf-categories.lagda.md [hunk 3]
```diff
@@ -65,34 +66,98 @@ module _
...
+    comp-hom-copresheaf-Large-Category X Y Y
+      ( id-hom-copresheaf-Large-Category Y)
+      ( f) ＝
+    f
+  left-unit-law-comp-hom-copresheaf-Large-Category X Y f =
+    left-unit-law-comp-hom-Large-Precategory
+      ( copresheaf-Large-Precategory)
+      { X = X}
+      { Y}
+      ( f)
```
Comment (fredrik-bakke) [inline src/category-theory/copresheaf-categories.lagda.md:144]:
```suggestion
  left-unit-law-comp-hom-copresheaf-Large-Category X Y =
    left-unit-law-comp-hom-Large-Precategory
      ( copresheaf-Large-Precategory)
      { X = X}
      { Y}
```

File: src/category-theory/copresheaf-categories.lagda.md [hunk 4]
```diff
@@ -65,34 +66,98 @@ module _
...
+    comp-hom-copresheaf-Large-Category X X Y
+      ( f)
+      ( id-hom-copresheaf-Large-Category X) ＝
+    f
+  right-unit-law-comp-hom-copresheaf-Large-Category X Y f =
+    right-unit-law-comp-hom-Large-Precategory
+      ( copresheaf-Large-Precategory)
+      { X = X}
+      { Y}
+      ( f)
```
Comment (fredrik-bakke) [inline src/category-theory/copresheaf-categories.lagda.md:160]:
```suggestion
  right-unit-law-comp-hom-copresheaf-Large-Category X Y =
    right-unit-law-comp-hom-Large-Precategory
      ( copresheaf-Large-Precategory)
      { X = X}
      { Y}
```

File: src/category-theory/presheaf-categories.lagda.md [hunk 5]
```diff
@@ -57,34 +59,96 @@ module _
...
-    right-unit-law-comp-hom-Large-Precategory presheaf-Large-Precategory
+  comp-hom-presheaf-Large-Category :
+    {l3 l4 l5 : Level}
+    (X : obj-presheaf-Large-Category l3)
+    (Y : obj-presheaf-Large-Category l4)
+    (Z : obj-presheaf-Large-Category l5) →
+    hom-presheaf-Large-Category Y Z → hom-presheaf-Large-Category X Y →
+    hom-presheaf-Large-Category X Z
+  comp-hom-presheaf-Large-Category X Y Z g f =
+    comp-hom-Large-Precategory presheaf-Large-Precategory {X = X} {Y} {Z} g f
```
Comment (fredrik-bakke) [inline src/category-theory/presheaf-categories.lagda.md:89]:
```suggestion
  comp-hom-presheaf-Large-Category X Y Z =
    comp-hom-Large-Precategory presheaf-Large-Precategory {X = X} {Y} {Z}
```

File: src/category-theory/presheaf-categories.lagda.md [hunk 6]
```diff
@@ -57,34 +59,96 @@ module _
...
+  associative-comp-hom-presheaf-Large-Category X Y Z W h g f =
+    associative-comp-hom-Large-Precategory
+      ( presheaf-Large-Precategory)
+      { X = X}
+      { Y}
+      { Z}
+      { W}
+      ( h)
+      ( g)
+      ( f)
```
Comment (fredrik-bakke) [inline src/category-theory/presheaf-categories.lagda.md:120]:
```suggestion
  associative-comp-hom-presheaf-Large-Category X Y Z W =
    associative-comp-hom-Large-Precategory
      ( presheaf-Large-Precategory)
      { X = X}
      { Y}
      { Z}
      { W}
```

File: src/category-theory/presheaf-categories.lagda.md [hunk 7]
```diff
@@ -57,34 +59,96 @@ module _
...
+    comp-hom-presheaf-Large-Category X Y Y
+      ( id-hom-presheaf-Large-Category Y)
+      ( f) ＝
+    f
+  left-unit-law-comp-hom-presheaf-Large-Category X Y f =
+    left-unit-law-comp-hom-Large-Precategory
+      ( presheaf-Large-Precategory)
+      { X = X}
+      { Y}
+      ( f)
```
Comment (fredrik-bakke) [inline src/category-theory/presheaf-categories.lagda.md:136]:
```suggestion
  left-unit-law-comp-hom-presheaf-Large-Category X Y =
    left-unit-law-comp-hom-Large-Precategory
      ( presheaf-Large-Precategory)
      { X = X}
      { Y}
```

File: src/category-theory/presheaf-categories.lagda.md [hunk 8]
```diff
@@ -57,34 +59,96 @@ module _
...
+    (f : hom-presheaf-Large-Category X Y) →
+    comp-hom-presheaf-Large-Category X X Y f
+      ( id-hom-presheaf-Large-Category X) ＝
+    f
+  right-unit-law-comp-hom-presheaf-Large-Category X Y f =
+    right-unit-law-comp-hom-Large-Precategory
+      ( presheaf-Large-Precategory)
+      { X = X}
+      { Y}
+      ( f)
```
Comment (fredrik-bakke) [inline src/category-theory/presheaf-categories.lagda.md:151]:
```suggestion
  right-unit-law-comp-hom-presheaf-Large-Category X Y =
    right-unit-law-comp-hom-Large-Precategory
      ( presheaf-Large-Precategory)
      { X = X}
      { Y}
```

**Comment by fredrik-bakke** [comment]
I wrote a response in your Abelianization PR about how we manage coauthorship. I think it is totally fine if we do not attribute coauthorship for merely writing review suggestions.

**Comment by fredrik-bakke** [comment]
(merged without attributing co-authorship to me)

**Comment by EgbertRijke** [comment]
Well, in this case the amount of suggestions was about the same as the amount of work I did to create this PR, so actually I made a conscious decision when I turned on the auto-merge with you as a coauthor. Anyway, I see that you already removed yourself before I could put you back in:)

**Comment by fredrik-bakke** [comment]
I don't think so. It was your idea to define this convention, and your initiative to fix the current issues. In some sense, it is inappropriate to give me credit, as that suggests that I have some intellectual ownership over the convention, which I don't currently feel I do.

Actually, I think we should reconsider our conventions for when we attribute coauthorship. I know I have been personally very sloppy on this point, but I don't want it to become a problem that people write and accept code suggestions, as this is a very efficient tool for reviewing code. Therefore it seems fitting to have a general rule never to attribute co-authorship for them.

**Comment by fredrik-bakke** [comment]
I'm sorry to step over you here btw. I just figured it would be easiest to demonstrate by example, and I felt some time pressure because of the auto-merge 😅

**Comment by EgbertRijke** [comment]
No worries! I'm impressed that you managed to catch this at 3am:)

**Comment by EgbertRijke** [comment]
> I don't think so. It was your idea to define this convention, and your initiative to fix the current issues. In some sense, it is inappropriate to give me credit, as that suggests that I have some intellectual ownership over the convention, which I don't currently feel I do.
> 
> Actually, I think we should reconsider our conventions for when we attribute coauthorship. I know I have been personally very sloppy on this point, but I don't want it to be a problem that people write and accept code suggestions, as this is a very efficient tool for reviewing code. Therefore it seems fitting to have a general rule never to attribute co-authorship for them.

My request actually applied only to that specific pull request, since I worked so hard for it, but yeah I think it is a good idea to establish a convention. 

In some cases it seemed appropriate to share credit, because both the original author and reviewer did a lot of work for it, and in some cases the work done by the original author clearly outweighs the work done by the reviewer. Sometimes the reviewer also does some of the thinking that goes into the pull request. This is common with refactoring PRs. So I can see it going both ways. I've generally been happy with how things have been going.

On the other hand, you are 100% right that with a clear convention people don't have to ask for being the sole author, like I did, because they might be uncomfortable doing so. So I'm in favor of a convention, and we should probably automate that convention if it is possible. Maybe we can borrow some of @VojtechStep braincells for this:)

**Pull request merged**